guides/keras_hub/rag_pipeline_with_keras_hub.py

```diff
@@ -0,0 +1,531 @@
+                    len(filtered_lines) == 1 or filtered_lines[-2] != ""
+                ):
+                    filtered_lines.append("")
+            skip_next = False
+
+    # Remove any empty lines and excessive whitespace
+    cleaned_text = "\n".join(
+        [l for l in filtered_lines if l.strip() or l == ""]
+    ).strip()
+
+    return cleaned_text
+
+
+"""
+## RAG Pipeline
+
+Implements the Retrieval-Augmented Generation (RAG) pipeline:
+
+- Extracts `features` from the `query image` and `database images`.
+- Finds the most `similar image` in the database.
+- Uses the `retrieved report` and the `query image` as input to the `Gemma3(1B) text` model to generate a new report.
+
+Returns the index of the matched image, the retrieved report, and the generated report.
+"""
+
+
+def rag_pipeline(query_img_path, db_image_paths, db_reports, vision_model, text_model):
+    """
+    Retrieval-Augmented Generation pipeline using vision model for retrieval and a compact text model for report generation.
+    Args:
+        query_img_path (str): Path to the query image
+        db_image_paths (list): List of database image paths
+        db_reports (list): List of database reports
+        vision_model: Vision model for feature extraction
+        text_model: Compact text model for report generation
+    Returns:
+        tuple: (best_idx, retrieved_report, generated_report)
+    """
+    # Extract features for the query image
+    query_features = extract_image_features(query_img_path, vision_model)
+    # Extract features for the database images
+    db_features = np.vstack(
+        [extract_image_features(p, vision_model) for p in db_image_paths]
+    )
+    # Ensure features are numpy arrays for similarity search
+    db_features_np = np.array(db_features)
+    query_features_np = np.array(query_features)
+    # Similarity search
+    similarity = np.dot(db_features_np, query_features_np.T).squeeze()
+    best_idx = np.argmax(similarity)
+    retrieved_report = db_reports[best_idx]
+    print(f"[RAG] Matched image index: {best_idx}")
+    print(f"[RAG] Matched image path: {db_image_paths[best_idx]}")
+    print(f"[RAG] Retrieved context/report:\n{retrieved_report}\n")
+    PROMPT_TEMPLATE = (
+        "Context:\n{context}\n\n"
+        "Based on the above radiology report and the provided brain MRI image, please:\n"
+        "1. Provide a diagnostic impression.\n"
+        "2. Explain the diagnostic reasoning.\n"
+        "3. Suggest possible treatment options.\n"
+        "Format your answer as a structured radiology report.\n"
+    )
+    prompt = PROMPT_TEMPLATE.format(context=retrieved_report)
+    # Generate report using the text model (text only, no image input)
+    output = text_model.generate(
+        {
+            "prompts": prompt,
+        }
+    )
+    cleaned_output = clean_generated_output(output, prompt)
+    return best_idx, retrieved_report, cleaned_output
+
+
+"""
+## Vision-Language Model (Direct Approach)
```

Comment:
Add a few sentences about - to compare the outputs of the RAG pipeline we just built. lets try how a generic VLM performs, etc ..
the guide should explain the flow of what and why